Approach.md

```diff
@@ -0,0 +1,102 @@
+
+## rrocker-cli:
+The CLI will need to be run once per command, so scheduling multiple tasks requires multiple invocations.
+
+There'll be zero command line switches and as such it'll be hardcoded to connect to `rrockerd` running on localhost.
+
+## rrocker-lib:
+No attention will be paid to backwards compatibility of the API, meaning no versioning or abstractions.
+
+Human friendliness of the API is a distant afterthought, as such full UUIDs will be required to interact with tasks.
+
+Resource limits will be exposed 1-to-1 with the underlying cgroups API leaving it up to the `rrocker-cli` to expose it in a human friendly way.
+# Security
+Security is naturally an important aspect of the implementation however as my experience with crypto is limited I'll attempt to follow best practices and use the default configuration of OpenSSL and rustTLS wherever applicable.
+In an actual production system you'd have a security expert decide the configuration based on a threat model.
+
+## Authentication
+Authentication will be done with mTLS according to the challenge rules.
+
+The following certificates will be created:
+- Root Certificate Authority (CA). This is our root of trust and should be stored on an air-gapped system that's only used to sign/revoke the server/client CAs.
+- Separate server and client CAs. These are used to sign/revoke server and client certs. Furthermore they're separated such that an infrastructure team is able to own their CA and deploy new servers independently. These CAs also introduce an indirection such that servers and clients don't need to know about all clients and servers but can simply verify an identity has been signed by the server/client CA.
+- Server_1 cert used by the server to auth with the clients
+- Client_1 and Client_2 certs used to demo user's only being able to see their own tasks
+- Admin_1 cert used to demo admin's being able to see all tasks
+- An untrusted CA + client cert to demo unauthorized client's can't connect
+
+Private keys will be generated with OpenSSL using the prime256v1 ECDH curve as recommended by mozilla: https://wiki.mozilla.org/Security/Server_Side_TLS#Modern_compatibility.
+
+## Authorization
+The authorization scheme will be super simple solution where an authenticated user either is an admin or regular user.
+A regular user will only be able to perform actions on it's own tasks while an admin can interact with **all** tasks.
+
+This distinction will be done by looking at the Common Name of the client certificate, `CN=client` for clients and `CN=admin` for admins.
+The public key of a cert will be used as the client id despite the drawback that a single client would have to share the same cert across multiple servers against best practices. 
+In a production environment you'd introduce another indirection such a user could have multiple certs all manage the same tasks and possibly some way to further segment these tasks into groups.
```

Comment:
Could you please point me towards a system that implements a model like described here. I'm not totally sure I understand / follow what the advice would be here for a production environment.

Reply:
The indirection I mention here could be implemented in two ways from the top of my head. 
One would be database backed where an authenticated request has it's public key looked up in a table that maps between public keys and client_id which is then used to return the appropriate set of tasks. This is similar to how github maps a multiple ssh keys to one github account.
Another approach could be to use **our** client CA to sign a subclient CA with which the customer could sign a certificate for each of their servers and then include the subclient CA in the certificate chain used for authentication. I'm not sure there are any production systems using this model as it's rather cumbersome for the customer to manage their own CA.

Reply:
Ok, I think I follow the line of argument. This is mostly about avoiding the copying of private keys to different hosts acting as a client that may want to present the same identity to the service. 

So the way this is often approached in mTLS is to not use the public key as the identity, but the subject field of the certificate. So using kubernetes as an example, you would issue certificates with something like `Subject: CN=kevin, O=cloud-dev O=security`. Kubernetes when it receives a certificate like this, will take the subject and read the CN as the user and the O fields as any groups. So kubernetes will interpret that Subject as user=kevin and groups=cloud-dev,security when it applies RBAC on what API requests I'm allowed to make based on both groups and principal. 

To uniquely identify an issued certificate, would be based on the Serial Number. So if you want to selectively disable a certificate, you would find the serial and add it to a revocation list, if you're using standard revocation, which has it's faults. For user identity, another common approach is to issue short lived certificates. So if the certificate is only valid for one hour, the expiry takes care of cancelling credentials that may leak.

Reply:
Yes that's indeed a much better approach than what I came up with. I've updated the design doc to reflect this change. Thanks!